### ISSUE TITLE
Creating datastore fails with custom schema as custom before_dataset_index is not called

### ISSUE THREAD

**Issue by stefina** [ckan/ckan#7926]
## CKAN version
2.10.1

## Describe the bug
Adding a datastore fails after CKAN-Upgrade from 2.8. when using a custom schema (multilingual)

### Steps to reproduce
- Apply custom dataset schema with multilingual values.
- load valid csv into datastore
- fails with `File "/srv/app/src/ckan/ckan/lib/search/index.py", line 47, in escape_xml_illegal_chars return _illegal_xml_chars_re.sub(replacement, val) TypeError('expected string or bytes-like object')`
- logging val in the function, our values are coming in like this: `{"fr": "Test Dataset", "de": "Test Dataset", "en": "Test Dataset", "it": "Test Dataset"}`

### Expected behavior
Our custom `before_dataset_index` should be called, so illegal xml characters can be replaced from the desired values ('Test Dataset' instead of multi-valued dict)

### Additional details
[This is where the re-indexing for the package is called](https://github.com/ckan/ckan/blob/8ddc2e800793d683c856b4a56f3d3f0ad3f1d168/ckanext/datastore/logic/action.py#L704) which is not calling the implemented `before_dataset_index`-interface from our custom plugin.
In comparison: When rebuilding a single dataset or the whole index, our custom function is called.

Is this behaviour intentional? With CKAN 2.8 we were able to load the same resource into the datastore.

**Comment by stefina** [comment]
After some more investigation it occured to me, that the before_dataset_index is indeed called, but only after the `escape_xml_illegal_chars`-function is called. I will try out, whether calling it earlier causes any other problems or whether it fixes this problems, but any input is appreciated.

**Comment by stefina** [comment]
@amercader @pdelboca thank you for assigning this issue. Maybe we can find a solution by discussing this?
I might be overlooking other options that we already have to fix this but I will try to desribe the problem more detailed here, maybe that will help me find a work-around or help us to find a solution in CKAN.

When a datastore is added, it needs to be added to the index. While indexing in general works for us, as we are using the `before_dataset_index`.

In order to make it possible to index the dataset anyways, we would have to normalize the dataset's title and to allow it to be consumed multilingual afterwards, we would have to change the title afterwards to a multilang dict again. As of now I can't think of a solution to do this without hacking around in CKAN core which we obviously want to avoid.

Can you think of a way, on how to find a work-around for this?

**Comment by amercader** [comment]
@stefina , just to check, are you using [ckanext-fluent](https://github.com/ckan/ckanext-fluent) to handle the multilingual fields?

**Comment by stefina** [comment]
@amercader yes, we do. @bellisk had the good idea to check the presets in use which I hope I get to do tomorrow, but if you have hints for this, they are very appreciated!

**Comment by amercader** [comment]
The fluent documentation discourages using it directly on core fields like `title` or `notes` to avoid issues like the one you are describing: https://github.com/ckan/ckanext-fluent#fluent_core_translated-fields

Perhaps you can adapt your schema to reflect this?

**Comment by stefina** [comment]
@amercader Our schema is unfortunately using those fields for years now amongst a few of our projects and changing that would have large side-effects regarding the shared schema.
So far we worked with a custom preset to be able to use this field with multiple languages, too.

@bellisk did some more digging and found the change, that prevents this to work for us:
This issue https://github.com/ckan/ckan/issues/6672 includes the following commit
https://github.com/ckan/ckan/commit/e1dde691fd12283209ccea39592c31e7013b25be#diff-4f2653bfd681793bf0b7ea57d10cefb6bcdf489bbd65dd71a8acb38e1f49854bR625 

In the function `set_datastore_active_flag` we used to get the package from solr directly, but now, calling `package_show` returns us the package directly from the database which calls the output validators on the package.
So, before we would get the title like this: `'{"de": "foo", "en": "foo", "fr": "foo", "it": "foo"}'` but since CKAN 2.10 we get the dict object.

We are considering whether to waive that fix in favour of getting our datastore to work makes sense, probably by adding a patch that reverses this change to our CKAN patches.
